fl4health/strategies/fedpca.py

```diff
@@ -0,0 +1,269 @@
+
+        is the SVD of B, then it turns out that U = A @ U',
+        where the columns of U' are the true principal components of the aggregated data,
+        and A is some block unitary matrix.
+
+        For the theoretical justification behind this procedure, see the paper
+        "A Distributed and Incremental SVD Algorithm for Agglomerative Data Analysis on Large Networks".
+
+        Args:
+            client_singular_vectors (NDArrays): Local PCs.
+            client_singular_values (NDArrays): Singular values corresponding to local PCs.
+
+        Returns:
+            Tuple[NDArray, NDArray]: merged PCs and corresponding singular values.
+        """
+        X = [U @ np.diag(S) for U, S in zip(client_singular_vectors, client_singular_values)]
+        svd_input = np.concatenate(X, axis=1)
+        new_singular_vectors, new_singular_values, _ = np.linalg.svd(svd_input, full_matrices=True)
+        return new_singular_vectors, new_singular_values
+
+    def merge_subspaces_qr(
+        self, client_singular_vectors: NDArrays, client_singular_values: NDArrays
+    ) -> Tuple[NDArray, NDArray]:
+        """
+        Produce the principal components (PCs) for all the data distributed across clients by merging the PCs
+        belonging to each local dataset.
+
+        Each clients sends a matrix whose columns are the local principal components to the server. The corresponding
+        singular values are also shared.
+
+        This implementation can be viewed as a more efficient approximation to
+        the SVD-based merging in that it does not require performing SVD on a large matrix.
+
+        Directly performing SVD does not take into account the following two observations, suggesting there are more
+        efficient algorithms for merging:
+            1. Each client's singular vectors are already orthonormal.
+            2. The right singular vectors do not need to be computed since
+            only the left singular vectors are returned as the merging result.
+
+        In contrast, the algorithm here performs a QR decomposition on the large data matrix, which
+        is more efficient than SVD, and SVD is only performed on a much smaller matrix.
+
+        Similarly to the SVD-based merging, it returns an approximation of the true principal components
+        of the aggregated data up to the multiplication of some block unitary matrix.
+
+        For the theoretical justification behind this approach, see the paper
+        "Subspace Tracking for Latent Semantic Analysis".
+
+        Args:
+            client_singular_vectors (NDArrays): Local PCs.
+            client_singular_values (NDArrays): Singular values corresponding to local PCs.
+
+        Returns:
+            Tuple[NDArray, NDArray]: merged PCs and corresponding singular values.
+        """
+        assert len(client_singular_values) >= 2
+        if len(client_singular_values) == 2:
+            singular_values_diagonal_matrices = [
```

Comment:
A more straightforward construction might be clearer given the length assertion. That is, 
```
U1, S1 = client_singular_vectors[0], np.diag(client_singular_values[0])
U2, S2 = client_singular_vectors[1], np.diag(client_singular_values[1])
```
and drop the list comprehension above.